script.js

```diff
@@ -0,0 +1,370 @@
+
+//****************************************************************
+//----- APPLY-----
+//****************************************************************
+
+const calf  = {
+  name:"caffee",
+  code:334
+};
+
+console.log(cow.getCode.apply(calf ,["baby"]));//baby334
+
+//****************************************************************
+//----- BIND-----
+//****************************************************************
+
+
+// call and apply will not make copy of function, It will only call the functiom
+// bind will make copy of function and we can call it
+
+const goat = {name:"Gii", code:101};
+
+const functionBind = sheep.getCode.bind(goat); //functionBind is brand new function for obj goat
+
+console.log(functionBind()); //101
+
+//****************************************************************
+//----- Constuctor functions-----
+//****************************************************************
+
+function Animal(animalID){
+  this.code = animalID;
+  this.getCode = function(){
+    console.log("Code: "+this.code);
+  };
+}
+
+const dog = new Animal(555);//constuctor function
+dog.getCode() //Code: 555
+
+//If we create 100000 objects of Animal,
+// getCode function will be repliated 100000 times
+//bz it get created in constuctor function
+//so here comes Prototype
+
+
+//****************************************************************
+//----- Prototype -----
+// you can create many objects of constuctor function but only one function will be created
+//****************************************************************
+
+
+function NewAnimal(id){
+  this.code = id;
+}
+
+NewAnimal.prototype.getCode = function(){
+  console.log("Code: " + this.code);
+}
+
+const cat = new NewAnimal(666);
+cat.getCode();// Code: 666
+
+//Expanding objects using prototype
+NewAnimal.prototype.appendPrefix = function(){
+  return " Batch: "+ this.code;
+}
+
+cat.appendPrefix(); // Batch: 666
+
+//****************************************************************
+//----- Array Methods-----
+//****************************************************************
+
+const valueArray = [1,2,3,4,5];
+const double = x => x * 2;
+
+// map
+console.log(valueArray.map(double)); //[2,4,6,8,10]
+
+const triple = x => x * 3;
+
+console.log(valueArray.map(triple));//[3,6,9,12,15]
+
+const greaterThan10 = x => x > 10;
+console.log([1,2,3,4,5,12,34,43].filter(greaterThan10));//[12,34,43]
+
+
+const animalData = [
+  {name:"coco", type:"dog", code:111},
+  {name:"meow", type:"cat", code:151},
+  {name:"baa", type:"sheep", code:211},
+  {name:"scooby", type:"dog", code:311}
+];
+
+//forEach
+animalData.forEach(animal => console.log(animal));
+
+//filter
+const getDogs= animalData.filter( animal => animal.type === "dog"); // {name:"coco", type:"dog", code:111} {name:"scooby", type:"dog", code:311}
+
+console.log(getDogs);
+
+//find - finds only first match and returns
+const getCodeGreaterThan200 = animalData.find(
+  animal => animal.code > 200
```

Comment:
You can extract this arrow function in a named function
const isAnimalCodeGreaterThan200 = animal => animal.code > 200
const animalsWithCodeGreaterThan200 = animalData.find(isAnimalCodeGreaterThan200);